### PR TITLE
Github Actions Workflow for OGC-API with Observations

### PR DESCRIPTION
**Pull request by webb-ben**
### Overview

This Branch and Pull Request were created to address the addition of a Github Actions test workflow that is able to demonstrate that, with our minimal test data, we can go end to end in wis2node.  

Given:

```
  - Test Data (CSVs in this iteration)
    - Metadata (station, discovery)
    - Configurations
```

Workflow (CLI):

```
  - Create discovery metadata
  - Publish discovery metadata as pygeoapi collection
  - Cache station metadata
    - Publish stations metadata as pygeoapi collection
    - Setup data pipeline
    - Transform data (the real workflow will be eventually driven with MQTT, but we need CLI workflow)
      - CSV -> BUFR && GeoJSON
      - Publish the BUFRs (and GeoJSONs) to a web folder for HTTP WAF access and PubSub (MQTT)
      - Publish the GeoJSONs into the pygeoapi backend
```

Successful workflow run should have:

```
  - BUFR file goes in the right directory?
  - GeoJSON file goes in the right directory?
  - Metadata gets published to pygeoapi backend?
  - Observations get published to pygeoapi backend?
  - Errors/edge case checking/handling
```

Thus far, this is what the branch has implemented when the test workflow run locally:

- [x] Create discovery metadata
- [x] Publish discovery metadata as pygeoapi collection
- [x] Cache station metadata
  - [x] Publish stations metadata as pygeoapi collection
  - [x] Setup data pipeline
  - [ ] Transform data (the real workflow will be eventually driven with MQTT, but we need CLI workflow)
    - [x] CSV -> BUFR && GeoJSON
    - [ ] Publish the BUFRs (and GeoJSONs) to a web folder for HTTP WAF access and PubSub (MQTT)
    - [x] Publish the GeoJSONs into the pygeoapi backend

### Where we are now (and issues)

- **Github Action VMs run a docker environment unlike what most users will be using**
  - Although the workflow runs on locally, the vm has trouble with running wis2node and sarracenia ([example workflow with this logged](https://github.com/webb-ben/wis2node/runs/4564215715?check_suite_focus=true)). This means that although the workflow doesn't fail, you will see that the pygeoapi, wis2node, and data-consumer/sarracenia containers are not running or restarting and failing. 
  
image

[(run that produced the picture)](https://github.com/wmo-im/wis2node/runs/4564024650?check_suite_focus=true)
  - At what point are we testing against the Github VM and at what point are we testing against the various ways someone might want to deploy wis2node?
  - The VM cannot build docker images from github repos, so I had to use docker images that I built separately for wis2node-ui and sarracenia.
- **What is the test workflow seeking to validate?**
  - This wis2node test workflow is caught between developmental steps of wis2node. Wis2node will be on PYPI, but with the current deployment, the only time wis2node is used, is as a docker container. It seems like we are on track for it to almost always be deployed this way. On the other hand this isn't actually testing against python 3.6, 3.8, or 3.9, just 3.8 which is what is installed by the dockerfile.
  - The `$WIS2NODE...` environment variables are not used in the current `make up && make build` deployment. Should they be mapped into the docker-compose file?
  - Do we need two separate workflows? A first one that builds out the tests without docker to test wis2node against different python versions. Then a second to verify that there is also docker compatibility and a valid end-to-end workflow.

### Specifics about each file (and things that might need to change before this branch can be merged)

- **`tests/load_obs_data.py`:** Reads a file or filesystem and combines all csvs with the like headers into one file, named `observations.csv` inside of the working directory. The headers are that of the first CSV read as well as X, Y, and Z columns retrieved from the station metadata.
- **`tests/Dockerfile`, `tests/build.sh`, and `tests/docker-compose.yml`:** Are used to transform and upload the `observations.csv` into a docker image of `kartoza/postgis`. This could be built out as a separate github action using docker. For minimal viable product, the files are in the tests folder.
- **`docker/pygeoapi/pygeoapi-config.yml`:** Added hardcoded mappinging fro the observations postgis provider. On local machines `localhost` usually needs to be replaced with `host.docker.internal`. Hypothetically you could run pygeoapi with the CSV provider and the output directly from `tests/load_obs_data.py`. That being said, the CSV provider does not have the necessary queryable options to run with wis2node-ui and faces scaling issues.
- **`docker/docker-compose.yml`:** Modifications were only made to run with github actions. Can be reverted once we figure out the other issues I hope.
- **`.github/workflows/tests.yml`:** Added to the workflow are the following steps:
  1. `setup test data ⚙️ PostgreSQL/PostGIS 📦`: Creation of the postgres backend for pygeoapi/observations.
  2. `build Wis2node ⚙️`: Initial build and run of wis2node with `make build` and `make up`.
  3. `generate wis2node collection`: Execution of wis2node scripts on test data. Run to create the station collection, discovery metadata, and observations bufr data.
  4. `run wis2node`: Re run wis2node and print running docker containers.

I am interested to hear others thoughts regarding the best steps going forward.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -22,9 +22,38 @@ jobs:
         pip3 install -r requirements.txt
```
Comment (tomkralidis) [inline .github/workflows/tests.yml:22]: Given these tests are bound to docker workflow, rename from `.github/workflows/tests.yml` to `.github/workflows/tests-docker.yml`

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -22,9 +22,38 @@ jobs:
...
+          && wis2node data observations process \
+            --discovery-metadata /app/tests/data/metadata/discovery/surface-weather-observations.yml \
+            --station-metadata /app/tests/data/metadata/station/0-454-2-AWSNAMITAMBO.json \
+            --mappings /app/tests/data/config/malawi_synop_bufr.json \
+            --geojson_mappings /app/tests/data/config/malawi_synop_json.geojson \
+            /app/tests/data/observations/Namitambo_SYNOP.csv \
+          && exit 0"
+    - name: run wis2node
+      run: |
+        sleep 10
```
Comment (tomkralidis) [inline .github/workflows/tests.yml:54]: Are we `sleep`-ing so as to let the data ingest complete?

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -15,11 +15,12 @@ services:
 
   wis2node-ui:
     container_name: wis2node-ui
-    build: https://github.com/wmo-im/wis2node-ui.git#main
+    image: webbben/wis2node-ui
```
Comment (tomkralidis) [inline docker/docker-compose.yml:18]: Shouldn't we point to the canonical repo?

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -15,11 +15,12 @@ services:
...
 
   pygeoapi:
     container_name: pygeoapi
     image: geopython/pygeoapi:latest
+    restart: always
```
Comment (tomkralidis) [inline docker/docker-compose.yml:23]: Rationale/impact?

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -40,7 +41,7 @@ services:
 
   data-consumer:
     container_name: data-consumer
-    build: https://github.com/MetPX/sarracenia.git#v03_wip
+    image: webbben/sarracenia
```
Comment (tomkralidis) [inline docker/docker-compose.yml:44]: Same here, shouldn't we point to the canonical repo?

File: tests/load_obs_data.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+###############################################################################
+
+import sys
```
Comment (tomkralidis) [inline tests/load_obs_data.py:22]: Alphabetize imports.

File: tests/load_obs_data.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+import os
+import csv
+import json
+
+station2geom = {}
+
+
+def station_to_geom(station_id=''):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:30]: add docstring and function annotation

File: tests/load_obs_data.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            fc = json.loads(file.read())
+            for f in fc['features']:
+                wsid = f['properties']['wigos_id']
+                geom = zip(('X', 'Y', 'Z'), f['geometry']['coordinates'])
+                station2geom[wsid] = {key: val for (key, val) in geom}
+    if station_id:
+        return station2geom[station_id]
+
+
+def walk_path(path):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:42]: add docstring and function annotation

File: tests/load_obs_data.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+# under the License.
+#
+###############################################################################
+
+import sys
+import os
+import csv
+import json
+
+station2geom = {}
```
Comment (tomkralidis) [inline tests/load_obs_data.py:27]: Could/should this be scoped locally in `station_to_geom` ?

File: tests/load_obs_data.py (hunk 7 above)
Comment (tomkralidis) [inline tests/load_obs_data.py:30]: Can we make `station_id: str = None`?

File: tests/load_obs_data.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    if station_id:
+        return station2geom[station_id]
+
+
+def walk_path(path):
+    """
+    Walks os directory path collecting all CSV files.
+
+    :param path: required, string. os directory.
+    :return: list. List of csv paths.
```
Comment (tomkralidis) [inline tests/load_obs_data.py:47]: `:returns:`

File: tests/load_obs_data.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    for root, _, files in os.walk(path, topdown=False):
+        for name in files:
+            if name == 'observations.csv':
+                continue
+            file_list.append(os.path.join(root, name))
+
+    return file_list
+
+
+def _handle_csvs(files, csv_out=[]):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:59]: 1. add docstring and function annotation
2. I wonder whether can have this as a locally scoped function within `handle_csv`?

File: tests/load_obs_data.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    Splits list of csv files into individual csv files.
+
+    :param files: required, string. URL to be shortened.
```
Comment (tomkralidis) [inline tests/load_obs_data.py:63]: add a `:returns:` def

File: tests/load_obs_data.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    else:
+        return parse_csv(file)
+
+
+def parse_csv(filename, ret_csv=[]):
+    """
+    Parse CSV file into csv.
+
+    :param filename: required, string. URL to be shortened.
+    :return: list. Parsed csv.
```
Comment (tomkralidis) [inline tests/load_obs_data.py:90]: `:returns:`

File: tests/load_obs_data.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        csv_reader = csv.reader(fp)
+        headers = [h.strip() for h in next(csv_reader)]
+        headers.extend(['X', 'Y', 'Z'])
+        if len(ret_csv) == 0:
+            ret_csv.append(headers)
+        for line in csv_reader:
+            station_id = line[headers.index('Station_ID')].strip().upper()
+            parsed_line = []
+            for h in ret_csv[0]:
+                if h == 'X' or h == 'Y' or h == 'Z':
```
Comment (tomkralidis) [inline tests/load_obs_data.py:104]: `if h in ['X', 'Y', 'Z']:`

File: tests/load_obs_data.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    for f in files:
+        print(f'Reading file: {os.path.basename(f)}\r', end='')
+        csv_out = parse_csv(f, csv_out)
+
+    print('\r\r')
+    return csv_out
+
+
+def handle_csv(file):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:73]: add function annotation

File: tests/load_obs_data.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    Parses and shortens CSV file.
+
+    :param file: required, string or list of strings.
```
Comment (tomkralidis) [inline tests/load_obs_data.py:77]: add `:returns:` def

File: tests/load_obs_data.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    if isinstance(file, list) and len(file) > 0:
+        return _handle_csvs(file)
+    else:
+        return parse_csv(file)
+
+
+def parse_csv(filename, ret_csv=[]):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:85]: add docstring and function annotation

File: tests/load_obs_data.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+                    parsed_line.append(line[headers.index(h)].strip())
+                else:
+                    parsed_line.append("")
+
+            ret_csv.append(parsed_line)
+
+    return ret_csv
+
+
+def write_csv(filename, content):
```
Comment (tomkralidis) [inline tests/load_obs_data.py:116]: add docstring and function annotation

**Review by tomkralidis — CHANGES_REQUESTED** [review]

File: tests/load_obs_data.py (hunk 9 above)
Comment (webb-ben) [reply tests/load_obs_data.py:27]: Its outside so the `data/metadata/station/stations.geojson` only needs to be read once.

File: tests/load_obs_data.py (hunk 11 above)
Comment (webb-ben) [reply tests/load_obs_data.py:59]: I think it can. The logic was pulled from something where I was using the function recursively. Should be all good though

File: docker/docker-compose.yml (hunk 3 above)
Comment (webb-ben) [reply docker/docker-compose.yml:18]: Yes they should. The builds can't be run from Github Repositories as you can see [from this error](https://github.com/webb-ben/wis2node/actions/runs/1584965243) in the workflow. Current workaround is Docker Image variants of these that I have been keeping up to date with their canonical repo.
image

File: docker/docker-compose.yml (hunk 5 above)
Comment (webb-ben) [reply docker/docker-compose.yml:43]: Same as ^

**Review by tomkralidis — APPROVED** [review]
